asynciterator.ts

```diff
@@ -1966,3 +1966,53 @@ type SourceExpression<T> =
 
 type InternalSource<T> =
   AsyncIterator<T> & { _destination: AsyncIterator<any> };
+
+/**
+ * @param source An AsyncIterator
+ * @returns The AsyncIterator if it is not empty, otherwise undefined
+ */
+export async function maybeIterator<T>(source: AsyncIterator<T>): Promise<null | AsyncIterator<T>> {
+  // Avoid creating a new iterator where possible
+  if ((source instanceof ArrayIterator || source instanceof BufferedIterator) && (source as any)._buffer?.length > 0)
+    return source;
+
+  if (source instanceof IntegerIterator && (source as any)._step >= 0 ? (source as any)._next <= (source as any)._last : (source as any)._next >= (source as any)._last)
+    return source;
+
+
+  let item;
+  do {
+    if ((item = source.read()) !== null)
+      return source.prepend([item]);
+    await awaitReadable(source);
```

Comment:
Yes, but the reattaching might get expensive. I would inline it here so setup and cleanup happen once.